### DISCUSSION TITLE
How to use nhost in Next.js route handler (app router)?

### DISCUSSION THREAD

**Discussion by stefan-girlich** [nhost/nhost#2932] [Q&A]
I want to create an nhost client in a Next.js 14 [route handler](https://nextjs.org/docs/app/building-your-application/routing/route-handlers).

There is a function [createServerSideClient](https://docs.nhost.io/reference/nextjs/create-server-side-client) from `@nhost/nextjs`, but its second parameter is of type [GetServerSidePropsContext](https://nextjs.org/docs/pages/api-reference/functions/get-server-side-props) which is only available when using the pages router, not with the app router.

It seems that `@nhost/nextjs` currently does not support route handlers. What is the best approach for creating an authenticated nhost client in a Node.js environment?

**Comment by stefan-girlich** [comment]
I have found the following solution for now:

### Client
Obtain current access token via [useAccessToken](https://docs.nhost.io/reference/react/use-access-token), then pass it via header.
```js
const response = await fetch(URL, {
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
    })
```


### Server (Node.js)
Extract token from header and pass it to feature. **Important**: Importing `NhostClient` from `@nhost/nextjs` fails on import. I needed to install `@nhost/nhost-js` and import `NhostClient` from there.

```js
const accessToken = req.headers.get('Authorization')?.replace('Bearer ', '')
  const nhost = new NhostClient({
    subdomain: process.env.NEXT_PUBLIC_NHOST_SUBDOMAIN,
    region: process.env.NEXT_PUBLIC_NHOST_REGION,
  })
  nhost.graphql.setAccessToken(accessToken) // or nhost.functions.setAccessToken or whatever functionality you want to use
```


_Update:_
The access token returned by useAccessToken() may be expired; there does not seem to be any auto-refreshing. `autoRefreshToken` ([docs](https://docs.nhost.io/reference/react/nhost-client)) seems to be false by default; will try setting to true explicitly

**Reply by dbarrosop** [reply]
Please, check this example, I think it might be useful to you: https://github.com/nhost/nhost/tree/main/examples/quickstarts/nextjs-server-components

**Reply by mifreudenthaler** [reply]
@dbarrosop this should be more prominent in the docs.
works perfectly.

**Reply by stefan-girlich** [reply]
@dbarrosop The example works great on the server-side; I can see via my browser dev tools that a cookie is set for `nhostSession` which is also read on the server. I'd expect that this cookie would also be used on the client-side via `NhostProvider` and a client-side `NhostClient` instance. However,  `useAuthenticationStatus()` would tell me that the user is not authenticated.

Am I missing something conceptually?

**Reply by stefan-girlich** [reply]
Found a solution; I just pass the session from the cookies to `NhostProvider`:

```ts
// in a top-level component, in my case layout.tsx
const $cookies = cookies()
const sessionCookieValue = $cookies.get(NHOST_SESSION_KEY)?.value || ''
let initialNhostSession: NhostSession | undefined = undefined
try {
  initialNhostSession = JSON.parse(atob(sessionCookieValue) || 'null')
} catch (error) {
  initialNhostSession = undefined
}

// elsewhere
<NhostProvider nhost={nhost} initial={initialNhostSession}>
```

The only downside that comes to mind is that the client running in the browser could refresh tokens and this change would not be presented in the cookie and thus not reach the Next.js server-side environment. Not sure if this could cause any issues.

**Reply by dbarrosop** [reply]
I guess you could use [onTokenChanged](https://docs.nhost.io/reference/javascript/auth/on-token-changed) to also update the session being sent to the backend.

**Reply by RDeluxe** [reply]
> Found a solution; I just pass the session from the cookies to `NhostProvider`:
> 
> ```ts
> // in a top-level component, in my case layout.tsx
> const $cookies = cookies()
> const sessionCookieValue = $cookies.get(NHOST_SESSION_KEY)?.value || ''
> let initialNhostSession: NhostSession | undefined = undefined
> try {
>   initialNhostSession = JSON.parse(atob(sessionCookieValue) || 'null')
> } catch (error) {
>   initialNhostSession = undefined
> }
> 
> // elsewhere
> <NhostProvider nhost={nhost} initial={initialNhostSession}>
> ```
> 
> The only downside that comes to mind is that the client running in the browser could refresh tokens and this change would not be presented in the cookie and thus not reach the Next.js server-side environment. Not sure if this could cause any issues.

Sorry to chime in, but I'm puzzled about how this would work.

`$cookies` requires to be called in a server component (layout.tsx) while `NhostProvider` is called in a client component. Importing a server component into a client component marks `layout.tsx` as a client component, and it then crashes.

My solution is to authenticate the user on both the server and the client independently (`nhost.auth.signIn` is missing some information, like `needsEmailVerification` anyway), but that sounds sketchy to me.

Am I missing some nhost/nextjs magic?

**Reply by RDeluxe** [reply]
Reading [this answer] (https://github.com/nhost/nhost/issues/2104#issuecomment-1778813011), am I to understand that there is no  built-in way to make a client-side NhostClient and server side client co-exist?

The docs are really lacking on this front, or I did not find anything. This issue seems to suggests that the user has to keep the server and client `nhostClient` in sync, but I'm not sure.

I also feel like the difference between the `nhost/react` and `nhost/next` should be more explicit, as importing the first one instead of the second in my client components will throw errors (telling me that the nhost client is not found in in the `NhostProvider`).

Do you have any example of using server+client side auth in sync, with a global NhostProvider?

**Reply by dbarrosop** [reply]
> $cookies requires to be called in a server component (layout.tsx) while NhostProvider is called in a client component

Not sure what the issue might be but conceptually this is what cookies are for so using them to transport the session from the client to the server should be the way to go.

> Reading [this answer] (https://github.com/nhost/nhost/issues/2104#issuecomment-1778813011), am I to understand that there is no built-in way to make a client-side NhostClient and server side client co-exist?

The way to synch clients should be via cookies as mentioned in a previous comment in this thread and as far as we know it works fine as we know of several users doing this.

> The docs are really lacking on this front, or I did not find anything. This issue seems to suggests that the user has to keep the server and client nhostClient in sync, but I'm not sure.

Unfortunately there are too many frameworks and too many ways of doing things so it is very hard to maintain and support every single one of them + all the different ways of doing things. If a particular use-case pops up enough times we may then try to write an official example. This is starting to come up enough times that we may want to look into writing one but we don't have a timeline for this.

> I also feel like the difference between the nhost/react and nhost/next should be more explicit,

We have them as separate client libraries in the [reference documentation](https://docs.nhost.io/reference/nextjs/nhost-client) and also as different [guides](https://docs.nhost.io/guides/quickstarts/nextjs) but if that's not enough and you have any suggestions on how to improve this it'd be much appreciated.

Thanks for your comments/feedback.
